### PR TITLE
Refine implicit priority change warnings

### LINKED ISSUE #20572: Incorrect warning about ambiguous implicit
## Compiler version

3.5.0-RC1-3.5.1-RC1-bin-20240613-61a15b2-NIGHTLY

## Minimized code

can't really minimize, but i hope the error message is enough
here is an approximation but for some reason it doesn't produce a warning:
https://scastie.scala-lang.org/OlegYch/io8Z28sZTC2qIUA3lqJtSw/36

## Output

instead i get a warning:
```
[warn] 124 |              "metal"      -> Money[CurrencyCode.Any](1, CurrencyCode.Gold),
[warn]     |                                                                           ^
[warn]     |Given search preference for play.api.libs.json.Writes[com.Money[com.CurrencyCode.CurrencyValue]] between alternatives (com.Money._MoneyFormat :
[warn]     |  [C <: com.CurrencyCode.Any]
[warn]     |    (implicit evidence$1: play.api.libs.json.Format[C]): play.api.libs.json.Format[com.Money[C]]
[warn]     |) and (play.api.libs.json.Writes.enumNameWrites :
[warn]     |  [E <: Enumeration](using e: ValueOf[E]): play.api.libs.json.Writes[e.value.Value]) will change
[warn]     |Current choice           : the first alternative
[warn]     |New choice from Scala 3.6: none - it's ambiguous
```
for some reason it tries to apply enumNameWrites from play, but it's only applicable for scala.Enumeration, not enumeratum

## Expectation
no warning

### LINKED ISSUE #21036: False positive warning for new implicit resolution rules in 3.5.0-RC3
## Compiler version

3.5.0-RC3
## Minimized code

```scala
//> using scala 3.5.0-RC3
// //> using options -source:3.6-migration

trait SameRuntime[A, B]
trait BSONWriter[T]
trait BSONHandler[T] extends BSONWriter[T]

opaque type Id = String
object Id:
  given SameRuntime[Id, String] = ???

given BSONHandler[String]                    = ???
given [T: BSONHandler]: BSONHandler[List[T]] = ???

given opaqueWriter[T, A](using
    rs: SameRuntime[T, A],
    writer: BSONWriter[A]
): BSONWriter[T] = ???

def write[A](a: A)(using writer: BSONWriter[A]): Unit = ???

val ids: List[Id] = ???

val x = write(ids)
```

## Output

```Compiling project (Scala 3.5.0-RC3, JVM (21))
[warn] ./ImplicitResolution2.scala:24:19
[warn] Given search preference for BSONWriter[List[Id]] between alternatives (given_BSONHandler_List :
[warn]   [T](implicit evidence$1: BSONHandler[T]): BSONHandler[List[T]]) and (opaqueWriter :
[warn]   [T, A](using rs: SameRuntime[T, A], writer: BSONWriter[A]): BSONWriter[T]) will change
[warn] Current choice           : the first alternative
[warn] New choice from Scala 3.6: none - it's ambiguous
[warn] val x = write(ids)
[warn]                   ^
Compiled project (Scala 3.5.0-RC3, JVM (21))
```

## Expectation

It shouldn't emit any warnings, as there is no implicit evidence for the second case. Also if we enabled `//> using options -source:3.6-migration` it works perfectly.

Related discussions: https://github.com/scala/scala3/pull/19300 and https://github.com/scala/scala3/issues/20484

### PR DESCRIPTION
**Pull request by odersky**
Fixes #21036
Fixes #20572

### CODE REVIEW

**Review by EugeneFlesselle — CHANGES_REQUESTED** [review]
Otherwise looks right to me 👍

File: compiler/src/dotty/tools/dotc/typer/Implicits.scala [hunk 1]
```diff
@@ -1443,8 +1447,8 @@ trait Implicits:
                       compareAlternatives(newCand, cand) > 0)
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Implicits.scala:1447]: I don't see how that follows. In `healAmbiguous` we either return a failure with `cand` in it or we switch to `newCand`. If either `cand` or `newCand` is in the final result, we do have a priority change.

But I think there's something else going on here. If compareAlternatives is -1 for in one scenario and 0 in the other it makes no difference and we should not even record a priority warning.

**Comment by EugeneFlesselle** [comment]
We now miss the following change:
```scala
trait A[T]
trait B[T] extends A[T]

given b[T]: B[T] = ???
given a[T]: A[T] = ???

val y = summon[A[String]] // changes from `b` in 3.5 to `a` 3.6 as expected but without warning in either
```

**Comment by EugeneFlesselle** [comment]
It seems we really do need to know whether the `cand` we are filtering out early would have been a `SearchSuccess` to decide whether or not to report a warning.

I guess an alternative solution would be to just disable the filtering of candidates for which `compareAlternatives` changes between the two modes.

File: compiler/src/dotty/tools/dotc/typer/Implicits.scala [hunk 2]
```diff
@@ -1319,19 +1321,24 @@ trait Implicits:
...
         def comp(using Context) = explore(compare(alt1.ref, alt2.ref, preferGeneral = true))
-        def warn(msg: Message) =
-          priorityChangeWarnings += ((alt1.ref, alt2.ref, msg))
         if alt1.ref eq alt2.ref then 0
         else if alt1.level != alt2.level then alt1.level - alt2.level
         else
           var cmp = comp(using searchContext())
           val sv = Feature.sourceVersion
-          if sv.stable == SourceVersion.`3.5` || sv == SourceVersion.`3.6-migration` then
+          if isWarnPriorityChangeVersion(sv) then
```
Comment (EugeneFlesselle) [inline compiler/src/dotty/tools/dotc/typer/Implicits.scala:1335]: I suppose `sv` is constant and `isWarnPriorityChangeVersion` could then be a val

**Review by EugeneFlesselle — APPROVED** [review]
This was tricky to do efficiently, but we got a good solution after all!

File: compiler/src/dotty/tools/dotc/typer/Implicits.scala (hunk 2 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Implicits.scala:1335]: No, I use it below as well in the assert that we originally commented out.

Comment (EugeneFlesselle) [reply compiler/src/dotty/tools/dotc/typer/Implicits.scala:1335]: Oh, I meant even `isWarnPriorityChangeVersion(Feature.sourceVersion)` would be constant between calls of `compareAlternatives`. But no big deal either way, this is still ready to merge for me

Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Implicits.scala:1335]: Yes, but be need sourceVersion as `sv` separately in compareAlternatives.

**Comment by odersky** [comment]
This should be part of 3.5.0 RC4.

**Pull request merged**